### PR TITLE
armvirt, malta: add support for kernel 6.1

### PR DESCRIPTION
**Pull request by guidosarducci**
This adds kernel 6.1 support for the malta and armvirt QEMU testing platforms.

First, add multiple kernel config settings missing from the generic config.
Next, create kernel 6.1 configs as copies of armvirt and malta 5.15 configs.
Then, update these armvirt and malta kernel 6.1 configs with new and changed symbols.
Finally, enable 6.1 as a testing kernel for armvirt and malta.

Build and run-tested on QEMU/malta-be32 and QEMU/malta-le64.
Build and run-tested on QEMU/armvirt32 and QEMU/armvirt64.

### CODE REVIEW

**Comment by guidosarducci** [comment]
@Ansuel Ah, forgot to ping you earlier for review on this 6.1 update further to your earlier work. Please have a look when you've a chance.

There are a lot of generic config updates in particular, including a little (expected) overlap with others' patches (e.g. https://github.com/openwrt/openwrt/pull/12703 from @robimarko or https://www.mail-archive.com/openwrt-devel@lists.openwrt.org/msg64526.html from @linusw ).

File: target/linux/generic/config-6.1
Comment (1715173329) [inline target/linux/generic/config-6.1:3766]: this should be handled by c07038da27cefa5a93e433909b9aca594386ddc1 already

Comment (guidosarducci) [reply target/linux/generic/config-6.1:3766]: Ah, yes, that was committed after my series so no surprise I missed it. Thanks, will drop that and rebase again.

**Comment by PolynomialDivision** [comment]
@guidosarducci Can you rebase on main? As you mentioned, your PR has overlap with other PRs that now got merged.

**Comment by guidosarducci** [comment]
@PolynomialDivision Thanks, just noticed the other PR was merged. Rebased now and looks OK. FYI, the CI tests have all passed before, but with all the current churn it seems new testing gets stalled...

File: target/linux/generic/config-6.1
Comment (PolynomialDivision) [inline target/linux/generic/config-6.1:5079]: Can you explain why you enable this? Just because it is the default value?

**Comment by 1715173329** [comment]
```
# CONFIG_DRM_ROCKCHIP is not set
# CONFIG_SND_AC97_POWER_SAVE is not set
# CONFIG_SND_SERIAL_GENERIC is not set
# CONFIG_SND_SOC_RK817 is not set
# CONFIG_SND_SOC_ROCKCHIP is not set
# CONFIG_TINYDRM_ILI9163 is not set
# CONFIG_TYPEC_DP_ALTMODE is not set
```

please also consider adding these configs to `generic/config-6.1`, they're in config-5.15 but somehow missing in 6.1

File: target/linux/generic/config-6.1
Comment (guidosarducci) [reply target/linux/generic/config-6.1:5079]: Previously, support for the `randomize_kstack_offset` was unconditionally compiled in for supported arches (arm64, ppc, x86, s390), so this changes nothing but just exposes the previous support for the feature.

Default usage is still determined by `RANDOMIZE_KSTACK_OFFSET_DEFAULT`, which is disabled both in kernel 5.15 and 6.1, with zero performance overhead. See https://lore.kernel.org/lkml/20220128114446.740575-1-elver@google.com/.

**Comment by guidosarducci** [comment]
@1715173329 How did you build this list exactly? Doesn't seem right...

> ```
> # CONFIG_SND_AC97_POWER_SAVE is not set
Above still exists in both 5.15 and 6.1 generic configs.
> ```
> # CONFIG_DRM_ROCKCHIP is not set
> # CONFIG_SND_SOC_RK817 is not set
> # CONFIG_SND_SOC_ROCKCHIP is not set
Above were removed in 6bdc6bb173ec8eaf67614b5e7e3890a8496cf5d9 as part of refresh by @Ansuel .
> ```
> # CONFIG_SND_SERIAL_GENERIC is not set
> # CONFIG_TINYDRM_ILI9163 is not set
> # CONFIG_TYPEC_DP_ALTMODE is not set
Above never existed in 5.15 or 6.1 generic configs
> ```
> 
> please also consider adding these configs to `generic/config-6.1`, they're in config-5.15 but somehow missing in 6.1

I prefer to keep changes in this PR related to enabling `armvirt` or `malta` per the title to avoid confusing the history. Regarding Rockchip above and the refresh, let's ask @Ansuel so he can fix, or maybe prefers I do it.

My experience with the kernel config refresh process is that it's poorly documented, poorly understood, and often does the wrong things. If I *have* to use it, I always review and try rationalize things afterwards. Maybe this a good opportunity to understand *why* the refresh dropped those config settings...

**Comment by guidosarducci** [comment]
Rebased and force-pushed again. The previous failed CI tests are all related to failing CDN downloads, which seems to be a recurring problem recently.

**Comment by 1715173329** [comment]
> How did you build this list exactly?

These options were exposed when I refreshing rockchip/config-6.1.

> Above still exists in both 5.15 and 6.1 generic configs.

I made a copy/paste error i guess ;)

> Above were removed in 6bdc6bb173ec8eaf67614b5e7e3890a8496cf5d9 as part of refresh by @Ansuel .

These options should be reserved tbh.

> Above never existed in 5.15 or 6.1 generic configs

Yep, they're new options introduced in 6.1.

> I prefer to keep changes in this PR related to enabling `armvirt` or `malta` per the title to avoid confusing the history.

Okay, I will create another PR for it once yours get merged. Thank you anyway!

**Comment by PolynomialDivision** [comment]
I'm fine with the PR. I am building everything again on my system and if works as expected I would merge.

**Comment by PolynomialDivision** [comment]
On a complete clean build, I get following error:
```
make[2]: Entering directory '/home/nick/openwrt/package/kernel/linux'
Package kmod-phylink is missing dependencies for the following libraries:
libphy.ko
make[2]: *** [modules/netdevices.mk:123: /home/nick/openwrt/bin/targets/armvirt/32/packages/kmod-phylink_6.1.29-1_arm_cortex-a15_neon-vfpv4.ipk] Error 1
make[2]: Leaving directory '/home/nick/openwrt/package/kernel/linux'
```

**Comment by guidosarducci** [comment]
@PolynomialDivision Hmm, thanks for the details. Are you using a `CONFIG_ALL` config? I've relied on the CI for feedback on full builds, but it just struck me that `armvirt` isn't a CI target for that. I wonder if we could add that in future for better coverage given how widespread it is?

Let me look at `armvirt` in the meantime...

**Comment by john-tho** [comment]
> On a complete clean build, I get following error:
> 
> ```
> make[2]: Entering directory '/home/nick/openwrt/package/kernel/linux'
> Package kmod-phylink is missing dependencies for the following libraries:
> libphy.ko
> make[2]: *** [modules/netdevices.mk:123: /home/nick/openwrt/bin/targets/armvirt/32/packages/kmod-phylink_6.1.29-1_arm_cortex-a15_neon-vfpv4.ipk] Error 1
> make[2]: Leaving directory '/home/nick/openwrt/package/kernel/linux'
> ```

This is accurate, and not specific to this PR. phylink does select libphy: https://elixir.bootlin.com/linux/latest/source/drivers/net/phy/Kconfig#L9 & https://github.com/torvalds/linux/commit/9e8d438e8ba43a38def2890a65a26917f2233a83
Our OpenWrt package should do the same. I would guess this has been missed because targets generally have their ethernet driver built in, and in that case any that select `PHYLIB` built in as part of the linux config would skip the OpenWrt module dependency check.
Something like:
```patch
diff --git a/package/kernel/linux/modules/netdevices.mk b/package/kernel/linux/modules/netdevices.mk
index 934ba3c4f2d2..bb62e5da9bcf 100644
--- a/package/kernel/linux/modules/netdevices.mk
+++ b/package/kernel/linux/modules/netdevices.mk
@@ -111,7 +111,9 @@ $(eval $(call KernelPackage,libphy))
 define KernelPackage/phylink
   SUBMENU:=$(NETWORK_DEVICES_MENU)
   TITLE:=Model for MAC to optional PHY connection
-  KCONFIG:=CONFIG_PHYLINK
+  KCONFIG:=CONFIG_PHYLINK \
+	   CONFIG_SWPHY=y
+  DEPENDS:=+kmod-libphy
   FILES:=$(LINUX_DIR)/drivers/net/phy/phylink.ko
   AUTOLOAD:=$(call AutoLoad,15,phylink,1)
 endef
```

---

Urgh. May be worth looking more into this:
On buildbot, kmod-phylink for armvirt64 is an empty package.
It is not built in the kmods stage: `WARNING: kmod-phylink is not available in the kernel config - generating empty package`

I _guess_, because `PHYLINK` does not have a Kconfig prompt, it only stays in kernel config if there is another that selects it. In a quick look in my build_dir, (`PHYLINK` is =m in [`.config.override`, and `.config.set`](https://github.com/openwrt/openwrt/blob/ef98dc3b3e07f529831c3ebe719d1a98bd43b568/include/kernel-defaults.mk#L114), but not the linux make processed `.config`
I will keep digging later. If I add Kconfig prompt text to `PHYLINK`, the config override works in linux config.

**Comment by guidosarducci** [comment]
> This is accurate, and not specific to this PR. phylink does select libphy: https://elixir.bootlin.com/linux/latest/source/drivers/net/phy/Kconfig#L9 & [torvalds/linux@9e8d438](https://github.com/torvalds/linux/commit/9e8d438e8ba43a38def2890a65a26917f2233a83)
> Our OpenWrt package should do the same. I would guess this has been missed because targets generally have their ethernet driver built in, and in that case any that select `PHYLIB` built in as part of the linux config would skip the OpenWrt module dependency check.

@john-tho  Thanks! That's a 2018 commit; I had done a pass over changes between 5.15 and 6.1 and found nothing -- as expected. I think you're right that many targets pull in `phylib` independently and avoid errors, but still surprised this goes back 5 years. I'm also trying to understand why `armvirt/32` wouldn't have encountered this before on kernel updates...

**Comment by guidosarducci** [comment]
@PolynomialDivision Building with `armvirt/32` locally and `CONFIG_ALL_KMODS=y`, I don't even get as far as your `kmod-phylink is missing dependencies` error. I run into multiple module compilation failures before that, e.g.

```
~/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armvirt_32/gl-mifi-mcu-1/module.c: In function 'gl_mifi_mcu_init': ~/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armvirt_32/gl-mifi-mcu-1/module.c:181:5: error: cast between incompatible function types from 'irqreturn_t (* (*)(unsigned int,  void *, struct pt_regs *))(int,  void *)' {aka 'enum irqreturn (* (*)(unsigned int,  void *, struct pt_regs *))(int,  void *)'} to 'irqreturn_t (*)(int,  void *)' {aka 'enum irqreturn (*)(int,  void *)'} [-Werror=cast-function-type]
  181 |     (irq_handler_t) handle_rx_start,
      |     ^
cc1: all warnings being treated as errors
```
or
```
make[8]: Entering directory '/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armvirt_32/linux-6.1.29'
  CC [M]  /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/coova-chilli-1.6/src/linux/xt_coova.o
../../coova-chilli-1.6/src/linux/xt_coova.c: In function 'coova_seq_open':
../../coova-chilli-1.6/src/linux/xt_coova.c:474:21: error: implicit declaration of function 'PDE_DATA'; did you mean 'NODE_DATA'? [-Werror=implicit-function-declaration]
  474 |         st->table = PDE_DATA(inode);
      |                     ^~~~~~~~
      |                     NODE_DATA
../../coova-chilli-1.6/src/linux/xt_coova.c:474:19: error: assignment to 'const struct coova_table *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  474 |         st->table = PDE_DATA(inode);
      |                   ^
../../coova-chilli-1.6/src/linux/xt_coova.c: In function 'coova_mt_proc_write':
../../coova-chilli-1.6/src/linux/xt_coova.c:486:33: error: initialization of 'struct coova_table *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  486 |         struct coova_table *t = PDE_DATA(file_inode(file));
      |                                 ^~~~~~~~
cc1: all warnings being treated as errors
```

On the other hand, building `armvirt/64` this way seems to work, but highlights some more missing generic configs. All very frustrating...

Please let me know what build configs you're using.

**Comment by guidosarducci** [comment]
Rebased and force-pushed, adding the following generic config settings needed building `armvirt/64`:
```
# CONFIG_DRM_PANEL_NEWVISION_NV3052C is not set
# CONFIG_DRM_PANEL_SAMSUNG_S6D27A1 is not set
# CONFIG_TINYDRM_ILI9163 is not set
```

**Comment by john-tho** [comment]
`gl-mifi-mcu`
> ```
> ~/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armvirt_32/gl-mifi-mcu-1/module.c: In function 'gl_mifi_mcu_init': ~/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armvirt_32/gl-mifi-mcu-1/module.c:181:5: error: cast between incompatible function types from 'irqreturn_t (* (*)(unsigned int,  void *, struct pt_regs *))(int,  void *)' {aka 'enum irqreturn (* (*)(unsigned int,  void *, struct pt_regs *))(int,  void *)'} to 'irqreturn_t (*)(int,  void *)' {aka 'enum irqreturn (*)(int,  void *)'} [-Werror=cast-function-type]
> ```
`coova-chilli`
> ```
> make[8]: Entering directory '/home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-armvirt_32/linux-6.1.29'
>   CC [M]  /home/kodidev/openwrt-project/build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/coova-chilli-1.6/src/linux/xt_coova.o
> ```

These are both from the packages feed, so they are easy to miss, and less important.

> Please let me know what build configs you're using.

I want to see that too! I guess you would need a kernel config package that selects `PHYLINK=m`, rather than just the depends from `SFP` because I cannot get that to stick with 5.15 armvirt64.

I don't think the `Package kmod-phylink is missing dependencies for the following libraries:
libphy.ko` issue saw should hold this PR up. It is a bit deeper, and long-existing.

**Comment by guidosarducci** [comment]
> These are both from the packages feed, so they are easy to miss, and less important.

@john-tho Yes, you're right. Those external packages are less important and only enabled when I set `CONFIG_ALL_KMODS=y`, but I've now found many of them and disabled them in my build. Nevertheless, I'll list them here so people are aware and can look into them later:
```
# CONFIG_PACKAGE_kmod-gl-mifi-mcu is not set
# CONFIG_PACKAGE_kmod-ipt-account is not set
# CONFIG_PACKAGE_kmod-ipt-chaos is not set
# CONFIG_PACKAGE_kmod-ipt-compat-xtables is not set
# CONFIG_PACKAGE_kmod-ipt-condition is not set
# CONFIG_PACKAGE_kmod-ipt-coova is not set
# CONFIG_PACKAGE_kmod-ipt-delude is not set
# CONFIG_PACKAGE_kmod-ipt-dhcpmac is not set
# CONFIG_PACKAGE_kmod-ipt-dnetmap is not set
# CONFIG_PACKAGE_kmod-ipt-fuzzy is not set
# CONFIG_PACKAGE_kmod-ipt-geoip is not set
# CONFIG_PACKAGE_kmod-ipt-iface is not set
# CONFIG_PACKAGE_kmod-ipt-ipmark is not set
# CONFIG_PACKAGE_kmod-ipt-ipp2p is not set
# CONFIG_PACKAGE_kmod-ipt-ipv4options is not set
# CONFIG_PACKAGE_kmod-ipt-length2 is not set
# CONFIG_PACKAGE_kmod-ipt-logmark is not set
# CONFIG_PACKAGE_kmod-ipt-lscan is not set
# CONFIG_PACKAGE_kmod-ipt-lua is not set
# CONFIG_PACKAGE_kmod-ipt-nathelper-rtsp is not set
# CONFIG_PACKAGE_kmod-ipt-proto is not set
# CONFIG_PACKAGE_kmod-ipt-psd is not set
# CONFIG_PACKAGE_kmod-ipt-quota2 is not set
# CONFIG_PACKAGE_kmod-ipt-sysrq is not set
# CONFIG_PACKAGE_kmod-ipt-tarpit is not set
# CONFIG_PACKAGE_kmod-pf-ring is not set
# CONFIG_PACKAGE_kmod-usb-serial-xr_usb_serial_common is not set
# CONFIG_PACKAGE_kmod-dahdi is not set
``` 

One of the core/non-feed wireless drivers (RLT8812-CT) also fails to build, and I've disabled it too since it's irrelevant to `malta` or `armvirt`. I do see there's newer code upstream which you've contributed to, so maybe we should update our package?
```
# CONFIG_PACKAGE_kmod-rtl8812au-ct is not set
```

I ran into another dependency problem with a core module package (not armvirt/malta specific):
```
removed '/home/kodidev/openwrt-project/bin/targets/armvirt/64/packages/kmod-rxrpc_6.1.29-1_aarch64_cortex-a53.ipk'
Package kmod-rxrpc is missing dependencies for the following libraries:
ip6_udp_tunnel.ko
udp_tunnel.ko
```
This looks straightforward from the history and I'll push a commit to fix it.

At this point I can successfully build `armvirt/64` and `armvirt/32` without error.

In neither case can I reproduce the error `Package kmod-phylink is missing dependencies for the following libraries: libphy.ko` seen by @PolynomialDivision , so I'd be comfortable having this PR merged after pushing the new commit and some review.

**Comment by PolynomialDivision** [comment]
I don't have the exact config anymore, since I had to clean it to build another target. But I'm very sure it looked like this because I did the same steps to reproduce it:
```
[nick@nicksbuildingvm openwrt]$ ./scripts/diffconfig.sh 
CONFIG_TARGET_armvirt=y
CONFIG_TARGET_armvirt_32=y
CONFIG_TARGET_armvirt_32_Default=y
CONFIG_ALL=y
CONFIG_ALL_KMODS=y
CONFIG_ALL_NONSHARED=y
CONFIG_BUILDBOT=y
CONFIG_COLLECT_KERNEL_DEBUG=y
CONFIG_HAS_BPF_TOOLCHAIN=y
CONFIG_IB=y
CONFIG_KERNEL_BUILD_DOMAIN="buildhost"
CONFIG_KERNEL_BUILD_USER="builder"
CONFIG_LINUX_6_1=y
CONFIG_MAKE_TOOLCHAIN=y
CONFIG_PACKAGE_bridger=m
CONFIG_PACKAGE_libxdp=m
CONFIG_PACKAGE_qosify=m
CONFIG_PACKAGE_xdp-filter=m
CONFIG_PACKAGE_xdp-loader=m
CONFIG_PACKAGE_xdpdump=m
CONFIG_REPRODUCIBLE_DEBUG_INFO=y
CONFIG_SDK=y
CONFIG_TESTING_KERNEL=y
CONFIG_UNETD_VXLAN_SUPPORT=y
CONFIG_USE_LLVM_BUILD=y
```

As you already discussed, I also think this PR is fine. The errors should be addressed in another issue/pr.

**Review by PolynomialDivision — APPROVED** [review]

**Comment by guidosarducci** [comment]
After the last few days, I'm stunned the CI tests completed, and all successfully! 😱

**Comment by Ansuel** [comment]
CI doesn't test kernel testing version :D

**Comment by guidosarducci** [comment]
Well, that's even _worse_, considering the number of failures, hangs, cancellations, and interruptions I've seen recently with CI. Just happy it completed and things seem more reliable now. :-)

On another note, as this is ready to merge should I be rebasing to simplify that?

**Comment by Ansuel** [comment]
@guidosarducci we already have a way to trigger runs based on label... I should expand that to test testing versions...

**Comment by guidosarducci** [comment]
@guidosarducci That sounds good. Would runs using test versions include building kmods/packages from external feeds e.g. packages? For test kernel testing, I'd expect packages to be restricted to core as the first step for a new kernel.

**Comment by Ansuel** [comment]
@guidosarducci for the sake of it i would limit it to kernel build and some kmods... the workflow already support that... but we are limited with runner so having stuff that runs for prolonged time is problematic.

**Comment by PolynomialDivision** [comment]
> On another note, as this is ready to merge should I be rebasing to simplify that?

No its fine. I will merge it. Thanks a lot!

**Pull request closed without merging**
